React Router v4

skipjack opened the pull request:
Still a WIP (see the relevant discussion in #99).

So far, I've removed the old `index.js` and `index.jsx` files and added a `/src` directory containing the new components, store, and (somewhat opinionated) stylesheet. I haven't updated the `/demo`, `package.json`, or `/test` directories much yet as I'd like to discuss a bit more. I'll have to dig into the build process some more as well. I do think there may be some opportunities to simplify a few things as we go but I'd like to make sure that's ok before jumping into anything. 

Once I've had a bit more time to jump into this I'll update with some thoughts. Until then, if anyone wants to dig into the `/src` files and try them out that'd be awesome. Definitely open to any feedback you all might have.

skipjack committed:
feat: rewrite library to be compatible with react-router v4

svenanders commented:
Thank you! I will be digging into it tomorrow! :)

lucascarvalho commented:
@svenanders hey, do you have any updates on this PR? :)

almirfilho commented:
It would be so nice to have this merged..
Plz bump!

skipjack commented:
@lucascarvalho @almirfilho what do you think of the proposed API for this? Did either of you try it out?

@svenanders definitely let me know your thoughts once you've had a chance to review. I think I'll have some time next week to help push this forward (e.g. add tests, update the demo, rewrite the readme, etc.).

gotdibbs commented:
Anything I could do to help move this along?

skipjack commented:
@MrLeebo have you had a chance to look at or play around with this? Just curious on your thoughts since I did talk this through with you in #99.

@gotdibbs and others -- I have some other updates/features once this starts moving along. And we would have to think about tests and documentation. I've been using this setup for a while now though with v4 and, while I've run into a few small quirks, it has worked really well overall.

@svenanders I'd love to move this along here but if you're unsure of the direction I'm happy to push it to a separate package as well.

MrLeebo commented:
@skipjack Honestly I haven't pursued this very much since that discussion. At work we opted to just keep using RRv3 for the foreseeable future since there doesn't seem to be a way to upgrade without refactoring nearly one hundred components in our app. I would be interested in seeing the [Friends demo](https://mrleebo.github.io/redux-breadcrumb-trail/#/friends/0) ported to this as a comparison, though.

svenanders commented:
I still think it's a good idea to do as we discussed earlier, that is, move the current master to a deprecated branch and use this PR as the base for a new master. I'm sorry I've been unable to work on this since before summer (feel really bad about this).

svenanders commented:
@skipjack The data object that is required as a props, what should be passed along here?

skipjack commented:
@svenanders it's basically a location descriptor with the addition of a `title` that can be any valid `PropTypes.node`. You can see it's used in the main component's `render` method. 

I actually have some time this week to help push this forward and as I said I have a few updates/fixes to push. To @MrLeebo's point, I will add a "friends" example as well as a README.

skipjack committed:
docs(readme): update readme to reflect v4 rework

Update the README to reflect the major changes in this package's
API. Note that the `/demo` is still a work in progress...

skipjack commented:
@svenanders @MrLeebo added the README, working on the `/demo` now.

skipjack committed:
refactor: rework components and styling

Update components based on other recent work and update
the styling to something a bit cleaner. Also remove postcss
syntax from `style.css` to keep things simple.

skipjack committed:
chore: update deps and babel config

Bring some dependencies up to date and modify the
base babel configuration. Some of this is prelim
work for the demo updates.

skipjack committed:
chore(demo): rewrite/redesign demo

Start updating the demo to work with the latest version of
the package. `npm init`ed the `/demo` directory so it can
contain it's own `package.json` and `dependencies`. Also
added `webpack-dev-server` for an easier dev workflow and
started to refactor the webpack configurations. The demo
still needs a little more work but it's getting close.

skipjack commented:
Ok started writing the demo -- though right now it only works on the `webpack-dev-server`. I still have some work to do on it which I'm hoping to finish up in the next couple days. @svenanders I feel a little bad tearing up the build processes but there were some things I couldn't figure out so I'm just trying to simplify the setup wherever I can. 

This is slowly and surely becoming a rewrite of the whole repo, so I hope you don't mind. I'd be happy to discuss and re-integrate any processes I intentionally or unintentionally remove.

svenanders commented:
That's cool. I thought the Makefile-system I employed was meaningful back in the day, but I don't mind it being refactored into something more up to date. There's also a lot of unnecessary complexity in the current demo folder that I don't mind removing either. So yeah, go ahead, and know I'm very happy and thankful that you're spending time on this!

skipjack committed:
chore(demo): finish work on demo app

Finished adding the other pages. Still need to update
the build process and add a readme but I think the
app itself is good to go.

skipjack committed:
chore(demo): update the build process

Rework the `webpack.prod.js` file and run a build.

skipjack committed:
docs(demo): add a readme

skipjack commented:
Ok pretty much finished the demo but there's still some more cleanup to do. Here's a quick screenshot:

![image](https://user-images.githubusercontent.com/8988697/30485292-ec7b46d8-99fa-11e7-9f86-efbec71d9c67.png)

> I thought the Makefile-system I employed was meaningful back in the day, but I don't mind it being refactored into something more up to date. There's also a lot of unnecessary complexity in the current demo folder that I don't mind removing either. So yeah, go ahead, and know I'm very happy and thankful that you're spending time on this!

Ok great, and yeah no problem at all! I'll let you know when I feel like the whole thing is in a good place -- and once it's there I could use some help with testing (like the actual automated testing not just reviewing/discussing the API).

skipjack committed:
refactor: rework and simplify build process

Removed a lot of the eslint config and removed the
`Makefile` in favor of `webpack` and npm `scripts`
usage across the board. Also added `standard-version`
and a `release` script.

skipjack committed:
chore: type check `children` and run a build

skipjack committed:
chore(demo): minor config changes/cleanup

skipjack commented:
Ok @svenanders et al I think this is ready for a review...

__Demo__

- `npm start`: Starts a `webpack-dev-server` for easy local development.
- `npm run build`: Creates a deployable demo in `/demo/dist`.

__Core__

- `npm run build`: Populates `/dist` with deployable files.
- `npm run release`: Utilizes [`standard-version`][1] to automate changelog generation and versioning.

Before running the above commands, you'll have to run an `npm install` in both the root directory and `/demo`. I think this could be avoided with some tweaks to the root configs if people feel 2 `package.json` files are overkill. Aside from that, I did strip a lot out so we could start with a simpler base and then discuss adding things back in if necessary. I would highly recommend we stick with `standard-version` for releasing as it makes life a lot easier (cleaner commit history, automated changelog, automated semver-compliant versioning).

Anyway, excited to hear feedback and hopefully push this forward. I am starting to run out of bandwidth to work on this so hopefully some others may be able to step in and help with certain things after we discuss what's needed.

[1]: https://github.com/conventional-changelog/standard-version

svenanders commented:
Wow! At first glance, this looks **fantastic**. The code looks clean and nice, and the important core functionality is all there. I'll be checking it out in depth when I have capacity later today, so I can understand everything that's going on and suggest some test cases.

skipjack commented:
> I'll be checking it out in depth when I have capacity later today, so I can understand everything that's going on and suggest some test cases. You've done more work on this than could ever be expected, and I'm ready to step up to maintain your excellent work.

Thanks! And yeah definitely let me know if I can clarify anything...

matpaul commented:
Any updates for this pull request ?

alexgurr commented:
+1 Waiting for V4 merge here!

svenanders commented:
Hi all! SORRY for taking forever. I've scoured and tested and I love this PR, and the only thing I've been waiting for is trying to add some content to this myself. HOWEVER, out of respect for @skipjack and all of you waiting, I can't delay this any longer, so here's what I'm doing: I'm moving the old code to a legacy branch, merging this wholesale and publishing this PR as the new main npm module (as 2.x), while keeping the old one available at 1.x

skipjack commented:
Awesome @svenanders, I'm interested to hear your thoughts on writing/adding tests! I guess the store could be tested in isolation but I don't have too much experience testing react components so I'm interested in what you think would be best.

svenanders commented:
Writing React component tests are pretty straightforward (at least nowadays), although I actually prefer to just the code logic and trust React to render it correctly. However, I think I'll go ahead and write up some tests using _enzyme_ and _chai-enzyme_ and publish is at as a PR for you to look at and comment on.

skipjack commented:
> Writing React component tests are pretty straightforward (at least nowadays), although I actually prefer to just the code logic and trust React to render it correctly. However, I think I'll go ahead and write up some tests using enzyme and chai-enzyme and publish is at as a PR for you to look at and comment on.

Sounds great.

Also, looking through the issues, I think a bunch will be resolved by this PR (e.g. #108) though I'm not sure if you'd like to keep them open to support react-router v3. The nice think about this solution though is that if we remove the `NavLink` dependency, and instead let it be passed via props, the new components would actually work for v3 as well.

svenanders commented:
Wow, that's worth exploring. I have a site that's using this component and is sort of stuck on React Router 3, so I'll definitely check out this solution to see if I can't use this new version rather than the old one by passing this down as a prop!
